PublicMapping/districtbuilder
# Fix sidebar flicker

## Overview

Fix sidebar flicker + various refactors

### Checklist


### Demo

![db_sidebar_flicker_fix](https://user-images.githubusercontent.com/2926237/91878060-cbdab880-ec4c-11ea-95c3-54e0aae09dd1.gif)

### Notes

This PR contains a lot of refactors, some of which turned out to not be strictly necessary but seemed valuable enough to include anyway (eg. making `getSidebarRows` into a component). The most notable example of this is using `reduceReducers` to combine district drawing state with project state. Having them be separate has required some kludgy workarounds in the past (see the now defunct `saveDistrictsDefinition`) in order to share data between them. I thought at one point that doing a single state update was important to avoid flicker, but that later turned out not to be true. In any case, this is a positive change (IMO) and will give us flexibility to avoid similar kludges in the future.

This also includes a large refactor of the `projectData` reducer which was important in terms of fixing this issue and also simplifies things considerably. The key change there was to add new S3/API functions to fetch all static data (see `fetchAllStaticData`) and fetch all updated project data (see `fetchProjectData`) this simplifies the `projectData` reducer considerably and should make it easier to manage state and avoid some of the unnecessary the intermediate states which trigger re-renders of components.

The other real, important change that had to be in this PR for it to work, in addition to some reworking of project state, is to make `assignGeounitsToDistrict` _not_ mutate the `DistrictsDefinition` (this was a mistake and it made everything super confusing to make sense of since a fundamental rule of Redux is that you don't mutate state). 

Another nice small change (IMO) was the addition of `destructureResource` to grab a key from inside a resource if it exists, which allows us to avoid a lot of repeated code in component functions (the way it's used in this PR is to grab it when wrangling Redux state and just pass the component what it needs directly to avoid dealing with resources).  

I also added a file for frontend-only types to avoid sharing them unnecessarily (see `src/client/types.d.ts`) and used various type aliases (eg. `DistrictsGeoJSON`, `UintArrays`) in more places on the frontend. It would be good to keep moving them over to that file in the future as appropriate.

## Testing Instructions
* Reassign geounits at different levels and ensure there is no flicker in the sidebar (only color of numbers change)
* Test that a loading state is shown in the sidebar header (spinner) while the request is pending

Closes #253
